### PR TITLE
bpo-26133: Dont unsubscribe signals in UNIX even loop on interpreter shutdown

### PR DESCRIPTION
**Pull request by asvetlov**
https://bugs.python.org/issue26133

### CODE REVIEW

File: Lib/asyncio/unix_events.py [hunk 1]
```diff
@@ -51,8 +51,12 @@ def __init__(self, selector=None):
...
     def close(self):
         super().close()
-        for sig in list(self._signal_handlers):
-            self.remove_signal_handler(sig)
+        if not sys.is_finalizing():
+            for sig in list(self._signal_handlers):
+                self.remove_signal_handler(sig)
+        else:
+            warinigs.warn(f"Closing the loop {self} on interpreter shutdown "
+                          f"stage, signal unsubsription is disabled")
```
Comment (1st1) [inline Lib/asyncio/unix_events.py]: stacklevel arg?

Comment (asvetlov) [reply Lib/asyncio/unix_events.py]: Sure! :)

File: Lib/asyncio/unix_events.py [hunk 2]
```diff
@@ -51,8 +51,13 @@ def __init__(self, selector=None):
...
         super().close()
-        for sig in list(self._signal_handlers):
-            self.remove_signal_handler(sig)
+        if not sys.is_finalizing():
+            for sig in list(self._signal_handlers):
+                self.remove_signal_handler(sig)
+        else:
+            warinigs.warn(f"Closing the loop {self} on interpreter shutdown "
+                          f"stage, signal unsubsription is disabled",
+                          stacklevel=2)
```
Comment (vstinner) [inline Lib/asyncio/unix_events.py]: I'm not sure that stacklevel=2 is correct. Maybe remove it?

For the warning category, I hesitate to use ResourceWarning and pass source=self. ResourceWarning are quiet by default. It's maybe better to not pollute stdout with such warning?

File: Lib/asyncio/unix_events.py [hunk 3]
```diff
@@ -51,8 +51,13 @@ def __init__(self, selector=None):
 
     def close(self):
         super().close()
-        for sig in list(self._signal_handlers):
-            self.remove_signal_handler(sig)
+        if not sys.is_finalizing():
+            for sig in list(self._signal_handlers):
+                self.remove_signal_handler(sig)
+        else:
+            warinigs.warn(f"Closing the loop {self} on interpreter shutdown "
```
Comment (vstinner) [inline Lib/asyncio/unix_events.py]: Please replace {self} with {self!r}. See for example BaseEventLoop.close() resource warning.

File: Misc/NEWS.d/next/Library/2017-12-21-11-08-42.bpo-26133.mt81QV.rst [hunk 4]
```diff
@@ -0,0 +1 @@
+Dont unsubscribe signals in asyncio UNIX event loop on interpreter shutdown.
```
Comment (vstinner) [inline Misc/NEWS.d/next/Library/2017-12-21-11-08-42.bpo-26133.mt81QV.rst]: It should be written "Don't", no?

**Comment by asvetlov** [comment]
Notes are fixed.
Not sure about stacklevel too. With level 2 traceback points on `__del__` most likely, with default it shows `.close()` explicitly.

File: Lib/asyncio/unix_events.py [hunk 5]
```diff
@@ -51,8 +51,13 @@ def __init__(self, selector=None):
...
         super().close()
-        for sig in list(self._signal_handlers):
-            self.remove_signal_handler(sig)
+        if not sys.is_finalizing():
+            for sig in list(self._signal_handlers):
+                self.remove_signal_handler(sig)
+        else:
+            warinigs.warn(f"Closing the loop {self!r} on interpreter shutdown "
+                          f"stage, signal unsubsription is disabled",
+                          ResourceWarning)
```
Comment (vstinner) [inline Lib/asyncio/unix_events.py]: For ResourceWarning, you can add source=self, so the warning will be logged with the traceback were the event loop was created, if traceback is enabled.

Comment (asvetlov) [reply Lib/asyncio/unix_events.py]: Done

**Review by vstinner — CHANGES_REQUESTED** [review]
LGTM.

It's hackish, but as I explained at https://bugs.python.org/issue26133#msg258429 it's hard to fix the root issue.

The hack is only used when the code doesn't explicitly close the event loop, which is a bug on itself. So it's ok to workaround https://bugs.python.org/issue26133 side effect.

**Review by vstinner — APPROVED** [review]

File: Lib/asyncio/unix_events.py [hunk 6]
```diff
@@ -51,8 +51,14 @@ def __init__(self, selector=None):
 
     def close(self):
         super().close()
-        for sig in list(self._signal_handlers):
-            self.remove_signal_handler(sig)
+        if not sys.is_finalizing():
```
Comment (1st1) [inline Lib/asyncio/unix_events.py:54]: What happens if someone calls `loop.close()` twice?  I guess `self._signal_handlers.clear()` is needed.

**Comment by 1st1** [comment]
> Not sure about stacklevel too. With level 2 traceback points on __del__ most likely, with default it shows .close() explicitly.

And what's the point of showing 'close()' explicitly? `__del__` gives you an idea that the warning is triggered by the GC.

**Comment by 1st1** [comment]
I'd appreciate if patches like this are merged with my approval.

**Comment by asvetlov** [comment]
@1st1 sorry

**Comment by 1st1** [comment]
NP, Andrew, it's not a big deal.  It's just that I have a few follow-up questions that I wish we discussed before the merge just in case.

File: Lib/asyncio/unix_events.py (hunk 6 above)
Comment (asvetlov) [reply Lib/asyncio/unix_events.py:54]: Double warning will be shown obviously.
But it is very rare situation: most likely the loop will be closed by `__del__`.
Even `atexit` callbacks are called when `sys.is_finalizing()` is `False`.

**Comment by asvetlov** [comment]
> And what's the point of showing 'close()' explicitly? `__del__` gives you an idea that the warning is triggered by the GC.

Stack trace points in `__del__` too.
Any `loop.close()` at interpreter finalizing stage is dangerous and should be fixed. Well, not dangerous but very non-accurate, I'd like to discourage this antipattern.

File: Lib/asyncio/unix_events.py (hunk 6 above)
Comment (1st1) [reply Lib/asyncio/unix_events.py:54]: Well, in this case I'd like to clear the list of signal handlers.

Comment (asvetlov) [reply Lib/asyncio/unix_events.py:54]: Ok, will make a postfix PR soon.

**Pull request merged**